unic/ucd/name/src/name.rs

```diff
@@ -9,38 +9,234 @@
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub enum Name {
+    Pieces(&'static [&'static str]),
+    HangulSyllable(char),
+    CJKUnifiedIdeograph(char),
+    TangutIdieograph(char),
+    NushuCharacter(char),
+    CJKCompatibilityIdeograph(char),
 }
 
 #[cfg_attr(feature = "clippy", allow(len_without_is_empty))]
 impl Name {
     pub fn of(ch: char) -> Option<Name> {
-        data::NAMES.find(ch).map(|pieces| Name { pieces })
+        match ch {
+            '\u{AC00}'...'\u{D7A3}' => Some(Name::HangulSyllable(ch)),
+            '\u{3400}'...'\u{4DB5}'
+            | '\u{4E00}'...'\u{9FEA}'
+            | '\u{20000}'...'\u{2A6D6}'
+            | '\u{2A700}'...'\u{2B734}'
+            | '\u{2B740}'...'\u{2B81D}'
+            | '\u{2B820}'...'\u{2CEA1}'
+            | '\u{2CEB0}'...'\u{2EBE0}' => Some(Name::CJKUnifiedIdeograph(ch)),
+            '\u{17000}'...'\u{187EC}' => Some(Name::TangutIdieograph(ch)),
+            '\u{1B170}'...'\u{1B2FB}' => Some(Name::NushuCharacter(ch)),
+            '\u{F900}'...'\u{FA6D}' | '\u{FA70}'...'\u{FAD9}' | '\u{2F800}'...'\u{2FA1D}' => {
+                Some(Name::CJKCompatibilityIdeograph(ch))
+            }
+            _ => data::NAMES.find(ch).map(|pieces| Name::Pieces(pieces)),
+        }
     }
 
     /// Length of the name in bytes.
     pub fn len(&self) -> usize {
-        // start with spaces
-        let mut len = self.pieces.len().saturating_sub(1);
-        for piece in self.pieces {
-            len += piece.len();
+        match self {
+            &Name::Pieces(pieces) => {
+                // start with spaces
+                let mut len = pieces.len().saturating_sub(1);
+                for piece in pieces {
+                    len += piece.len();
+                }
+                len
+            }
+            &Name::HangulSyllable(ch) => {
+                let mut len = PREFIX_HANGUL_SYLLABLE.len();
+                // FIXME: use decomposed jamo short names instead
+                len += self.num_of_hex_digits(ch as u32);
+                len
+            }
+            &Name::CJKUnifiedIdeograph(ch) => {
+                let mut len = PREFIX_CJK_UNIFIED_IDEOGRAPH.len();
+                len += self.num_of_hex_digits(ch as u32);
+                len
+            }
+            &Name::TangutIdieograph(ch) => {
+                let mut len = PREFIX_TANGUT_IDEOGRAPH.len();
+                len += self.num_of_hex_digits(ch as u32);
+                len
+            }
+            &Name::NushuCharacter(ch) => {
+                let mut len = PREFIX_NUSHU_CHARACTER.len();
+                len += self.num_of_hex_digits(ch as u32);
+                len
+            }
+            &Name::CJKCompatibilityIdeograph(ch) => {
+                let mut len = PREFIX_CJK_COMPATIBILITY_IDEOGRAPH.len();
+                len += self.num_of_hex_digits(ch as u32);
+                len
+            }
+        }
+    }
+
+    fn num_of_hex_digits(&self, code_point: u32) -> usize {
```

Comment:
You can write this as:

```rust
fn number_of_hex_digits(ch: char) -> usize {
    (32 - u32::leading_zeros(ch as u32) as usize + 3) / 4
}
```

Please note that it should be a static method, since `&self` is not needed in the impl. (`clippy` would have caught this, I think.)

Reply:
(I knew there existed a clever solution but I couldn't think of it)

Reply:
You'll get to master leading-zero/trailing-zero kinds of questions after school, in job interviews! ;)

Reply:
Oh, btw, @eyeplum, either make this a macro, or `#[inline=always]` it. There's no reason for it not being inline.

Reply:
Here's the `libcore` implementation of (still unstable) `char::escape_unicode`, which uses the same trick, but a bit more readable:

https://github.com/rust-lang/rust/blob/ddaab61101e48f0a58ce176b5a6aff11ae36c986/src/libcore/char.rs#L416-L432

For more fun like this, see this list: https://github.com/rust-lang/rust/search?utf8=%E2%9C%93&q=leading_zeros&type=

And, deep down, this is the impl: https://github.com/rust-lang/rust/blob/a538fe7ce715c7bd27e2e05329c3d857b9ad92af/src/libcore/num/mod.rs#L1357

Which calls LLVM's `ctlz`: https://llvm.org/docs/LangRef.html#llvm-ctlz-intrinsic

Reply:
Fixed, thank you!